### PR TITLE
Add wallet abstraction

### PR DESCRIPTION
**Pull request by KingGorrin**
Implementation for feature https://github.com/hypercore-one/znn_sdk_dart/issues/2

Some changes have been introduced since issue #2 

The additional changes include:

- `WalletManager`, `Wallet` and `WalletDefinition` interfaces

These interfaces enable the complete decoupling of the `KeyFile` specific wallet implementation. The same interfaces have been used in the [Zenon SDK for .NET](https://github.com/hypercore-one/znn_sdk_csharp/tree/wallet) library to implement a separate [ledger wallet](https://github.com/hypercore-one/znn_sdk_csharp/tree/ledger) library.

Using the wallet abstraction, the individual wallet libraries can be used in combination with each other, as demonstrated in the ledger implementation of the [Zenon CLI for .NET](https://github.com/hypercore-one/znn_cli_csharp/tree/bigint-ledger).

**Compatibility**

The changes are 98% compatible. Breaking changes include:

- The `address` property of the `KeyPair` class has been replaced by the async `getAddress()` method.
- The `saveKeyStore` method of the `KeyStoreManager` class now returns a `WalletDefinition` instead of a `File` class.

**Future possibilities**

Separate the file based wallet implementation from the main library.

### CODE REVIEW

**Comment by KingGorrin** [comment]
Tested the changes on latest [syrius](https://github.com/hypercore-one/syrius) build. No changes required, 100% backwards compatible.

File: lib/src/wallet/keypair.dart [hunk 1]
```diff
@@ -18,7 +19,7 @@ class KeyPair {
     return privateKey;
   }
 
-  Future<List<int>> getPublicKey() async {
+  Future<List<int>> getPublicKey([bool verify = false]) async {
```
Comment (vilkris4) [inline lib/src/wallet/keypair.dart]: What's the purpose of this positional parameter?

File: lib/src/wallet/keypair.dart [hunk 2]
```diff
@@ -32,7 +33,7 @@ class KeyPair {
   }
 
   Future<List<int>> sign(List<int> message) async {
-    return Crypto.sign(message, privateKey, (await getPublicKey()));
+    return Crypto.sign(Crypto.digest(message), privateKey, (await getPublicKey()));
```
Comment (vilkris4) [inline lib/src/wallet/keypair.dart]: Why this change?

Comment (KingGorrin) [reply lib/src/wallet/keypair.dart]: See the last paragraph in the https://github.com/hypercore-one/znn_sdk_dart/issues/2

An alternative is to pass a transaction interface containing both the source and the hash data as properties?

File: lib/src/wallet/keypair.dart (hunk 1 above)
Comment (KingGorrin) [reply lib/src/wallet/keypair.dart]: I added this for the ledger impl. In case the retrival of the public key needs to be verified by the user. It's specific to the ledger impl. Maybe it is better to move it out of the interface, into the specific ledger class impl.

File: lib/src/wallet/keypair.dart (hunk 2 above)
Comment (vilkris4) [reply lib/src/wallet/keypair.dart]: I may be missing something here, but why can't the caller just perform the hashing if needed?

Does this change work if you want to sign an arbitrary message?

File: lib/src/utils/block.dart [hunk 3]
```diff
@@ -163,36 +163,35 @@ class BlockUtils {
...
   static Future<bool> _setHashAndSignature(
-      AccountBlockTemplate transaction, KeyPair currentKeyPair) async {
-    transaction.hash = BlockUtils.getTransactionHash(transaction);
-    var transSig =
-        await BlockUtils._getTransactionSignature(currentKeyPair, transaction);
-    transaction.signature = transSig;
+      AccountBlockTemplate transaction, Wallet wallet) async {
+    var source = BlockUtils.getTransactionBytes(transaction);
+    transaction.hash = Hash.digest(source);
+    transaction.signature = await wallet.sign(source);
```
Comment (vilkris4) [inline lib/src/utils/block.dart]: `transaction.signature = await wallet.sign(transaction.hash.getBytes()!);`
Would this work if the `sign` method is left unchanged?

File: lib/src/wallet/keypair.dart (hunk 1 above)
Comment (vilkris4) [reply lib/src/wallet/keypair.dart]: I think it would make sense to move it away from this interface since it's more of a UI level feature requiring input from the user.

File: lib/src/wallet/keypair.dart (hunk 2 above)
Comment (KingGorrin) [reply lib/src/wallet/keypair.dart]: How does the caller know when the hashing is needed?

The ledger impl. needs the raw tx data. This information is used to display the `type`, `amount` and `toAddress` to the user for confirmation. After confirmation it calculates the hash from the raw tx and signs it.

This is different from the file based impl. which directly accepts and signs the hash.

In both cases it's the hash that is signed, but extra metadata is needed for confirmation in case of the ledger impl.

File: lib/src/utils/block.dart (hunk 3 above)
Comment (KingGorrin) [reply lib/src/utils/block.dart]: This is how it worked before the change. The hashing was done outside the sign method.

Using this method, the ledger impl. can only show the hash to the user for confirmation, because it does not have access to the raw tx data.

File: lib/src/wallet/keypair.dart (hunk 2 above)
Comment (KingGorrin) [reply lib/src/wallet/keypair.dart]: The old impl. was a low level function that signs the data. In case of the ledger impl. we need a method that is specific for signing a transaction. Therefore it's probably better to introduce a separate method `signTx` which accepts a tx and a generic one `signMsg` which just signs the data passed to it.

Comment (vilkris4) [reply lib/src/wallet/keypair.dart]: Yeah I think a separate method for signing a transaction makes sense.

Comment (KingGorrin) [reply lib/src/wallet/keypair.dart]: Something like this?

``` dart
/// Signs an arbitrary message.
Future<List<int>> sign(List<int> message);

/// Signs the hash of the transaction.
Future<List<int>> signTx(AccountBlockTemplate tx);
```
With the impl. being.
``` dart
Future<List<int>> sign(List<int> message) async {
  return Crypto.sign(message, privateKey, (await getPublicKey()));
}

Future<List<int>> signTx(AccountBlockTemplate tx) async {
  return Crypto.sign(
      Crypto.digest(tx.hash.getBytes()!), privateKey, (await getPublicKey()));
}
```
This will give other implementations of the interface full control over the transaction information.

**Comment by KingGorrin** [comment]
@georgezgeorgez mentioned he already did something similar in the go-zdk.

https://github.com/ignition-pillar/go-zdk/blob/master/wallet/interface.go

I will incorporate his changes and use the same interface name.

File: lib/src/wallet/keypair.dart (hunk 2 above)
Comment (vilkris4) [reply lib/src/wallet/keypair.dart]: Lgtm

File: lib/src/utils/block.dart [hunk 4]
```diff
@@ -163,36 +163,35 @@ class BlockUtils {
...
 
   static Future<bool> _setHashAndSignature(
-      AccountBlockTemplate transaction, KeyPair currentKeyPair) async {
-    transaction.hash = BlockUtils.getTransactionHash(transaction);
-    var transSig =
-        await BlockUtils._getTransactionSignature(currentKeyPair, transaction);
-    transaction.signature = transSig;
+      AccountBlockTemplate transaction, Signer signer) async {
+    var source = BlockUtils.getTransactionBytes(transaction);
+    transaction.hash = Hash.digest(source);
```
Comment (vilkris4) [inline lib/src/utils/block.dart]:
```suggestion
    transaction.hash = BlockUtils.getTransactionHash(transaction);
```
This could be restored to how it was originally. And I think the `BlockUtils.getTransactionBytes` is now an unused function?

Comment (KingGorrin) [reply lib/src/utils/block.dart]: I agree, and not really. The `getTransactionHash` method uses it.

`return Hash.digest(getTransactionBytes(transaction));`

Comment (vilkris4) [reply lib/src/utils/block.dart]: Yes because it was changed to work that way in this PR. But currently that change serves no purpose, since `getTransactionBytes` isn't needed anywhere else.

Obviously it doesn't really matter if it's changed to that but from a principles point of view, I prefer keeping changes as minimal as possible compared to the target branch.

In any case, this PR looks good so I'm approving it.

**Review by vilkris4 — APPROVED** [review]

File: lib/src/utils/block.dart (hunk 4 above)
Comment (KingGorrin) [reply lib/src/utils/block.dart]: I get what you're saying, but while the SDK doesn't use the method directly, it's actually quite convenient to keep them separate so that the ledger implementation can re-use it.

Although one might argue that you should isolate both implementations. In that case we would have to copy the method in the ledger impl.

Comment (vilkris4) [reply lib/src/utils/block.dart]: Okay I didn't realize the ledger implementation uses it. All good then.

**Review by vilkris4 — APPROVED** [review]

**Comment by KingGorrin** [comment]
I did one final pass on the wallet abstraction. I wasn't really happy with the `Signer` name and wanted to include an interface for the manager. Although I don't really like it, I've kept the variable names of the `Zenon` class intact for backwards compatibility.

With this addition the file store wallet implementation can be completely decoupled and replaced.

Also I've noticed that the dart tests fail because of something I've mentioned ealier in the bigint PR here https://github.com/zenon-network/znn_sdk_dart/pull/9#issuecomment-1584473382
The output does not equal the input.

**Review by vilkris4 — APPROVED** [review]
The changes to the manager seem to make sense. Maybe we create a separate ticket for renaming the variables in the `Zenon` class and refactor the clients at the same time.

**Comment by sol-znn** [comment]
> The changes to the manager seem to make sense. Maybe we create a separate ticket for renaming the variables in the `Zenon` class and refactor the clients at the same time.

I'm not sure where we've landed on this decision.

I propose we push through HTLC and BigInt upstream, then we can push these wallet changes.
Or do we want to include the wallet abstraction now?

**Comment by vilkris4** [comment]
I'd prefer getting HTLC and BigInt resolved first before we integrate these changes.

**Comment by KingGorrin** [comment]
I agree, we can do these changes afterwards.

**Comment by KingGorrin** [comment]
I'm planning to refactor this PR and prepare it for a new review.

**Comment by KingGorrin** [comment]
Updated PR description
